### PR TITLE
Allow to directly set the expireDate on a new (link)share

### PR DESCRIPTION
**Pull request by rullzer**
As requested by @PVince81 in #14808 and again in https://github.com/owncloud/core/pull/18631#issuecomment-135834038 ;)

Now it is possible to directly set the expiration date on shares.

TODO:
- [ ] Update documentations

### CODE REVIEW

**Comment by karlitschek** [comment]
sweeet :+1:

File: apps/files_sharing/api/local.php [hunk 1]
```diff
@@ -538,6 +552,30 @@ public static function deleteShare($params) {
...
+	 * If not throw an exception
+	 *
+	 * @param string $expireDate
+	 *
+	 * @throws \Exception
+	 * @return \DateTime
+	 */
+	private static function parseDate($expireDate) {
+		if (preg_match('/^\d{4}-\d{2}-\d{2}$/', $expireDate) === 0) {
+			throw new \Exception();
```
Comment (PVince81) [inline apps/files_sharing/api/local.php]: Add exception message "Invalid date" ?

File: apps/files_sharing/api/local.php [hunk 2]
```diff
@@ -286,6 +287,14 @@ public static function createShare($params) {
 				// read, create, update (7) if public upload is enabled or
 				// read (1) if public upload is disabled
 				$permissions = $publicUpload === 'true' ? 7 : 1;
+
+				// Get the expiration date
+				try {
+					$expirationDate = isset($_POST['expireDate']) ? self::parseDate($_POST['expireDate']) : null;
+				} catch (\Exception $e) {
+					return new \OC_OCS_Result(null, 404, 'Invalid Date');
```
Comment (PVince81) [inline apps/files_sharing/api/local.php]: Append message "format must be YYYY-DD-MM" ?

File: apps/files_sharing/api/local.php [hunk 3]
```diff
@@ -538,6 +552,30 @@ public static function deleteShare($params) {
...
+	 */
+	private static function parseDate($expireDate) {
+		if (preg_match('/^\d{4}-\d{2}-\d{2}$/', $expireDate) === 0) {
+			throw new \Exception();
+		}
+
+		$date = new \DateTime($expireDate);
+
+		if ($date === false) {
+			throw new \Exception();
```
Comment (PVince81) [inline apps/files_sharing/api/local.php]: Same here

File: lib/private/share/share.php [hunk 4]
```diff
@@ -649,6 +649,15 @@ public static function shareItem($itemType, $itemSource, $shareType, $shareWith,
 			$permissions = (int)$permissions & ~\OCP\Constants::PERMISSION_DELETE;
 		}
 
+		//Validate expirationDate
+		if ($expirationDate !== null) {
+			try {
+				$expirationDate = self::validateExpireDate($expirationDate->format('Y-m-d'), time(), $itemType, $itemSource);
```
Comment (PVince81) [inline lib/private/share/share.php]: can you add a comment to explain why we're passing `time()` here ?

**Comment by PVince81** [comment]
Works :+1:, tested with this pyocclient branch: https://github.com/owncloud/pyocclient/pull/120

**Comment by PVince81** [comment]
@rullzer can you fix the comments/strings, then feel free to merge

**Comment by SergioBertolinSG** [comment]
I don't find good to be able to put expiration dates before the current date.

**Comment by PVince81** [comment]
Raised doc ticket here: https://github.com/owncloud/documentation/issues/1615

**Comment by PVince81** [comment]
Now I wonder whether we should backport this.

Anyway, API consumers (like pyocclient) will have to work with older OC servers that don't have this API implemented.

**Comment by rullzer** [comment]
@PVince81 yeah not trivial. But clients will need to handle it properly anyway.

Since if sopmebody is still on 8.0.3 is should also work. So I do not think this has to be backported. Clients can check on version number. So that is probably simpler.

**Pull request merged**